man/ConicProj.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "CONICPROJ 1"
.TH CONICPROJ 1 2025-11-05 "GeographicLib 2.7" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
ConicProj \-\- perform conic projections
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBConicProj\fR ( \fB\-c\fR | \fB\-a\fR ) \fIlat1\fR \fIlat2\fR
[ \fB\-l\fR \fIlon0\fR ] [ \fB\-k\fR \fIk1\fR ] [ \fB\-r\fR ]
[ \fB\-e\fR \fIa\fR \fIf\fR ] [ \fB\-w\fR ] [ \fB\-p\fR \fIprec\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
Perform one of two conic projections geodesics.  Convert geodetic
coordinates to either Lambert conformal conic or Albers equal area
coordinates.  The standard latitudes \fIlat1\fR and \fIlat2\fR are specified
by that the \fB\-c\fR option (for Lambert conformal conic) or the \fB\-a\fR
option (for Albers equal area).  At least one of these options must be
given (the last one given is used).  Specify \fIlat1\fR = \fIlat2\fR, to
obtain the case with a single standard parallel.  The central meridian
is given by \fIlon0\fR.  The longitude of origin is given by the latitude
of minimum (azimuthal) scale for Lambert conformal conic (Albers equal
area).  The (azimuthal) scale on the standard parallels is \fIk1\fR.
.PP
Geodetic coordinates are provided on standard input as a set of lines
containing (blank separated) \fIlatitude\fR and \fIlongitude\fR (decimal
degrees or degrees, minutes, seconds);  for details on the allowed
formats for latitude and longitude, see the \f(CW\*(C`GEOGRAPHIC COORDINATES\*(C'\fR
section of \fBGeoConvert\fR\|(1).  For each set of geodetic coordinates, the
corresponding projected easting, \fIx\fR, and northing, \fIy\fR, (meters) are
printed on standard output together with the meridian convergence
\&\fIgamma\fR (degrees) and (azimuthal) scale \fIk\fR.  For Albers equal area,
the radial scale is 1/\fIk\fR.  The meridian convergence is the bearing of
the \fIy\fR axis measured clockwise from true north.
.PP
Special cases of the Lambert conformal projection are the Mercator
projection (the standard latitudes equal and opposite) and the polar
stereographic projection (both standard latitudes correspond to the same
pole).  Special cases of the Albers equal area projection are the
cylindrical equal area projection (the standard latitudes equal and
opposite), the Lambert azimuthal equal area projection (both standard
latitude corresponds to the same pole), and the Lambert equal area conic
projection (one standard parallel is at a pole).
.SH OPTIONS
.IX Header "OPTIONS"
.IP "\fB\-c\fR \fIlat1\fR \fIlat2\fR" 4
.IX Item "-c lat1 lat2"
use the Lambert conformal conic projection with standard parallels
\&\fIlat1\fR and \fIlat2\fR.
.IP "\fB\-a\fR \fIlat1\fR \fIlat2\fR" 4
.IX Item "-a lat1 lat2"
use the Albers equal area projection with standard parallels \fIlat1\fR and
\&\fIlat2\fR.
.IP "\fB\-l\fR \fIlon0\fR" 4
.IX Item "-l lon0"
specify the longitude of origin \fIlon0\fR (degrees, default 0).
.IP "\fB\-k\fR \fIk1\fR" 4
.IX Item "-k k1"
specify the (azimuthal) scale \fIk1\fR on the standard parallels (default
1).
.IP \fB\-r\fR 4
.IX Item "-r"
perform the reverse projection.  \fIx\fR and \fIy\fR are given on standard
input and each line of standard output gives \fIlatitude\fR, \fIlongitude\fR,
\&\fIgamma\fR, and \fIk\fR.
.IP "\fB\-e\fR \fIa\fR \fIf\fR" 4
.IX Item "-e a f"
specify the ellipsoid via the equatorial radius, \fIa\fR and
the flattening, \fIf\fR.  Setting \fIf\fR = 0 results in a sphere.  Specify
\&\fIf\fR < 0 for a prolate ellipsoid.  A simple fraction, e.g., 1/297,
is allowed for \fIf\fR.  By default, the WGS84 ellipsoid is used, \fIa\fR =
6378137 m, \fIf\fR = 1/298.257223563.
.IP \fB\-w\fR 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
on input and output, longitude precedes latitude (except that, on input,
this can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR,
\&\fIW\fR).
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 6).  \fIprec\fR is the number
of digits after the decimal point for lengths (in meters).  For
latitudes and longitudes (in degrees), the number of digits after the
decimal point is \fIprec\fR + 5.  For the convergence (in degrees) and
scale, the number of digits after the decimal point is \fIprec\fR + 6.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., "#" or "//").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP \fB\-\-version\fR 4
.IX Item "--version"
print version and exit.
.IP \fB\-h\fR 4
.IX Item "-h"
print usage and exit.
.IP \fB\-\-help\fR 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of "\-" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of "\-" stands for standard output.
.SH EXAMPLES
.IX Header "EXAMPLES"
.Vb 4
\&   echo 39.95N 75.17W | ConicProj \-c 40d58 39d56 \-l 77d45W
\&   => 220445 \-52372 1.67 1.0
\&   echo 220445 \-52372 | ConicProj \-c 40d58 39d56 \-l 77d45W \-r
\&   => 39.95 \-75.17 1.67 1.0
.Ve
.SH ERRORS
.IX Header "ERRORS"
An illegal line of input will print an error message to standard output
beginning with \f(CW\*(C`ERROR:\*(C'\fR and causes \fBConicProj\fR to return an exit
code of 1.  However, an error does not cause \fBConicProj\fR to
terminate; following lines will be converted.
.SH AUTHOR
.IX Header "AUTHOR"
\&\fBConicProj\fR was written by Charles Karney.
.SH HISTORY
.IX Header "HISTORY"
\&\fBConicProj\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in version 1.9.


NEWS
A reverse chronological list of changes to GeographicLib

For more information, see

    https://geographiclib.sourceforge.io/C++/doc/

The current version of the library is 2.7.

Changes between 2.7 (released 2025-11-06) and 2.6 versions:

   * Fix two bugs in classes for triaxial ellipsoids:
     + Triaxial::Geodesic3::Inverse sometimes threw an exception when
       both endpoints were on the major principal ellipse for very
       eccentric ellipsoids a > 2 c.
     + Triaxial::Cartesian3::carttocart2 sometimes returned the wrong
       result in the interior of the ellipsoid with Z = 0.

   * Cite the paper covering some of the triaxial classes Jacobi's
     solution for geodesics on a triaxial ellipsoid.

   * Minor documentation fixes.

Changes between 2.6 (released 2025-09-30) and 2.5.2 versions:

   * Support for triaxial ellipsoids. This is all placed in the
     GeographicLib::Triaxial namespace with the header files in the
     GeographicLib/Triaxial directory. This provides
     + Triaxial::Ellipsoid3 supports basic conversions of ellipsoidal
       coordinates.
     + Triaxial::GeodesicLine3 and Triaxial::Geodesic3 solve the direct
       and inverse geodesic problems on a triaxial ellipsoid. This
       implements Jacobi's solution for the direct problem. The inverse
       problem extends the method for solving the inverse problem on a
       biaxial ellipsoid.
     + Triaxial::Cartesian3 performs conversions between different
       systems of triaxial coordinates.
     + Triaxial::Conformal3 implements Jacobi's conformal projection for
       a triaxial ellipsoid. This allow a triaxial ellipsoid to be
       conformally mapped to a sphere or to any other ellipsoid.
     + Geod3Solve is a command-line utility offering access to
       Triaxial::Geodesic3.
     + Cart3Convert is a command-line utility offering access to
       Triaxial::Cartesian3.
     + Conformal3Proj is a command-line utility offering access to
       Triaxial::Conformal3.
     + Constants::Triaxial_Earth_a, Constants::Triaxial_Earth_b,
       Constants::Triaxial_Earth_c, and Constants::Triaxial_Earth_lon0
       specify a simple triaxial model for the Earth.
     + experimental::TriaxialGeodesicODE is provided to illustrate the
       solution of the direct geodesic problem by integrating the
       ordinary differential equations for the geodesic. Because this
       depends on Boost, this is not part of the GeographicLib library.
     + experimental/Geod3ODE.cpp a utility for doing direct geodesic
       calculations using experimental::TriaxialGeodesicODE.

   * To support the triaxial routines, the following classes have been
     added:
     + AngleT provides an accurate representation of angles of any
       magnitude. This might replace AuxAngle at some point.
     + Trigfun supports representing periodic functions as Fourier
       series with automatic selection of the number of terms to
       use. This offers a tiny subset of the functionality of the trig
       series in Chebfun. This might replace DST at some point.

   * C++17 is now required. Minimum version of Visual Studio supported
     is Visual Studio 15 2017.

   * Add Math::clamp.

   * Because experimental::TriaxialGeodesicODE depends on Boost, the
     CMake option USE_BOOST_FOR_EXAMPLES has been renamed
     USE_BOOST. This will cause find_package(Boost) to be called, but
     does not affect how the library or the tools are built.

Changes between 2.5.2 (released 2025-08-22) and 2.5.1 versions:

   * Fix potential buffer overflow in DMS.  Upgrading from earlier
     versions is strongly recommended.  If upgrading is not possible,
     avoid processing untrusted or malformed inputs and run programs
     linked to the library with restricted privileges.  Thanks to
     Rosario Matteo Grammatico for finding this vulnerability.

Changes between 2.5.1 (released 2025-08-20) and 2.5 versions:

   * Fix bug in geodesic inverse problem with very prolate
     ellipsoids. For f < -2, a meridional path was incorrectly
     returned in some cases for two points on the same meridian. This
     affected both the series and exact solutions. (Or course, the
     series solution is inaccurate for ellipsoids this eccentric.)

   * Minor tweak to Math::atan2d to improve accuracy.

   * Allow GEOGRAPHICLIB_PRECISION > 5 to allow use of mpreal with a
     fixed precision.

   * Document that cmake option -D BINDIR=OFF removes tools from the
     default all target for make.

   * WARNING: C++17 compliant compiler will be required for the next
     release. Support for builds with Visual Studio 14 (2015) will
     then stop.

Changes between 2.5 (released 2024-12-28) and 2.4 versions:

   * New magnetic models provided

     + World Magnetic Model 2025, wmm2025; this covers the period
       2025-2030.  This is now the model returned by
       MagneticModel::DefaultMagneticName and is default magnetic
       model for MagneticField (replacing wmm2020 which is only valid
       thru the end of 2024).
     + World Magnetic Model High Resolution 2025, wmmhr2025; this is a
       high-resolution version of wmm2025 including terms up to degree
       133 (as opposed to 12 for wmm2025).
     + International Geomagnetic Reference Field (14th generation),
       igrf14, which approximates the main magnetic field of the earth
       for the period 1900-2030.

   * Update CMake's cmake_minimum_required to 3.17.0 (to facilitate
     integration with Qt projects).

   * For GEOGRAPHICLIB_PRECISION > 2, readarray treats doubles as
     approximations to exact decimal constants.

Changes between 2.4 (released 2024-07-14) and 2.3 versions:

   * Math::sind, Math::sincosd, etc., now return accurate results for
     arguments = 30d, 45d, 60d, ...

   * Add Math::hypot3.

   * Add EllipticFunction::am.

   * Make constructor for AuxAngle explicit.

   * Include notes for plate carree areas.

   * C++14 compliant compiler required (this requirement tracks the

tools/ConicProj.cpp
/**
 * \file ConicProj.cpp
 * \brief Command line utility for conical projections
 *
 * Copyright (c) Charles Karney (2009-2017) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * http://geographiclib.sourceforge.net/
 *
 * See the <a href="ConicProj.1.html">man page</a> for usage information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/LambertConformalConic.hpp>
#include <GeographicLib/AlbersEqualArea.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions and potentially
// uninitialized local variables
#  pragma warning (disable: 4127 4701)
#endif

#include "ConicProj.usage"

int main(int argc, const char* const argv[]) {
  try {
    using namespace GeographicLib;
    typedef Math::real real;
    Utility::set_digits();
    bool lcc = false, albers = false, reverse = false, longfirst = false;
    real lat1 = 0, lat2 = 0, lon0 = 0, k1 = 1;
    real
      a = Constants::WGS84_a(),
      f = Constants::WGS84_f();
    int prec = 6;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';';

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-r")
        reverse = true;
      else if (arg == "-c" || arg == "-a") {
        lcc = arg == "-c";
        albers = arg == "-a";
        if (m + 2 >= argc) return usage(1, true);
        try {
          for (int i = 0; i < 2; ++i) {
            DMS::flag ind;
            (i ? lat2 : lat1) = DMS::Decode(std::string(argv[++m]), ind);
            if (ind == DMS::LONGITUDE)
              throw GeographicErr("Bad hemisphere");
          }
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-l") {
        if (++m == argc) return usage(1, true);
        try {
          DMS::flag ind;
          lon0 = DMS::Decode(std::string(argv[m]), ind);
          if (ind == DMS::LATITUDE)
            throw GeographicErr("Bad hemisphere");
          lon0 = Math::AngNormalize(lon0);
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-k") {
        if (++m == argc) return usage(1, true);
        try {
          k1 = Utility::val<real>(std::string(argv[m]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-e") {
        if (m + 2 >= argc) return usage(1, true);
        try {
          a = Utility::val<real>(std::string(argv[m + 1]));
          f = Utility::fract<real>(std::string(argv[m + 2]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        m += 2;
      } else if (arg == "-w")
        longfirst = !longfirst;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    if (!(lcc || albers)) {
      std::cerr << "Must specify \"-c lat1 lat2\" or "
                << "\"-a lat1 lat2\"\n";
      return 1;
    }

    const LambertConformalConic lproj =
      lcc ? LambertConformalConic(a, f, lat1, lat2, k1)
      : LambertConformalConic(1, 0, 0, 0, 1);
    const AlbersEqualArea aproj =
      albers ? AlbersEqualArea(a, f, lat1, lat2, k1)
      : AlbersEqualArea(1, 0, 0, 0, 1);

    // Max precision = 10: 0.1 nm in distance, 10^-15 deg (= 0.11 nm),
    // 10^-11 sec (= 0.3 nm).
    prec = std::min(10 + Math::extra_digits(), std::max(0, prec));
    std::string s, eol, stra, strb, strc;
    std::istringstream str;
    int retval = 0;
    while (std::getline(*input, s)) {
      try {
        eol = "\n";
        if (!cdelim.empty()) {
          std::string::size_type m = s.find(cdelim);
          if (m != std::string::npos) {
            eol = " " + s.substr(m) + "\n";
            s = s.substr(0, m);
          }
        }
        str.clear(); str.str(s);
        real lat, lon, x, y, gamma, k;
        if (!(str >> stra >> strb))
          throw GeographicErr("Incomplete input: " + s);
        if (reverse) {
          x = Utility::val<real>(stra);
          y = Utility::val<real>(strb);
        } else
          DMS::DecodeLatLon(stra, strb, lat, lon, longfirst);
        if (str >> strc)
          throw GeographicErr("Extraneous input: " + strc);
        if (reverse) {
          if (lcc)
            lproj.Reverse(lon0, x, y, lat, lon, gamma, k);
          else
            aproj.Reverse(lon0, x, y, lat, lon, gamma, k);
          *output << Utility::str(longfirst ? lon : lat, prec + 5) << " "
                  << Utility::str(longfirst ? lat : lon, prec + 5) << " "
                  << Utility::str(gamma, prec + 6) << " "
                  << Utility::str(k, prec + 6) << eol;
        } else {
          if (lcc)
            lproj.Forward(lon0, lat, lon, x, y, gamma, k);
          else
            aproj.Forward(lon0, lat, lon, x, y, gamma, k);
          *output << Utility::str(x, prec) << " "
                  << Utility::str(y, prec) << " "
                  << Utility::str(gamma, prec + 6) << " "
                  << Utility::str(k, prec + 6) << eol;
        }
      }
      catch (const std::exception& e) {
        *output << "ERROR: " << e.what() << "\n";
        retval = 1;
      }
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
